numpy/core/src/multiarray/scalarapi.c

```diff
@@ -314,54 +322,48 @@ PyArray_FromScalar(PyObject *scalar, PyArray_Descr *outcode)
                 NULL, (PyObject *)scalar);
     }
 
-    PyArrayObject *r = (PyArrayObject *)PyArray_NewFromDescr(&PyArray_Type,
-            typecode,
-            0, NULL,
-            NULL, NULL, 0, NULL);
+    if (outcode == NULL) {
+        outcode = typecode;
+        Py_INCREF(outcode);
+    }
+
+    PyArrayObject *r = (PyArrayObject *)PyArray_NewFromDescr_int(
+            &PyArray_Type, outcode, 0, NULL, NULL, NULL,
+            0, NULL, NULL, 0, 1);
     if (r == NULL) {
-        Py_XDECREF(outcode);
         return NULL;
     }
-    /* the dtype used by the array may be different to the one requested */
-    typecode = PyArray_DESCR(r);
-    if (PyDataType_FLAGCHK(typecode, NPY_USE_SETITEM)) {
-        if (typecode->f->setitem(scalar, PyArray_DATA(r), r) < 0) {
+    if (PyArray_DESCR(r) != outcode) {
+        /*
+         * Sanity check, only subarray dtypes should be modified, and there
+         * should be no subarray scalars (they would be arrays).
+         */
+        PyErr_SetString(PyExc_RuntimeError,
+                "dtype changed while creating an array from a scalar. This "
+                "should not be possible. Please notify the NumPy developers "
+                "you see this message");
+        Py_DECREF(r);
+        return NULL;
+    }
+    if (outcode == typecode) {
+        /*
+         * Fast path using setitem directly, since this is the matching,
+         * scalar for sure.
+         */
+        assert(PyObject_TypeCheck(scalar, NPY_DTYPE(outcode)->scalar_type));
```

Comment:
Hopefully only for void scalars, but subclasses can reach this code...

Reply:
rather than use an assert which will only work in debug mode, perhaps do a proper check-and-raise?

Reply:
Hmmm, its a fairly hot path, although possibly not very hot on the grand scheme of things.  I think our scalar API should only be used for subclasses of `np.generic` which are usually also registered as a dtype.  So not sure, I don't think there is a change here though, previously `scalar_value` would have crashed hard if it was not a correct `np.generic` subclass, I think.